dbms/src/Flash/Mpp/AsyncRequestHandler.h

```diff
@@ -0,0 +1,389 @@
+        read_packet_index = 0;
+        received_packet_index = 0;
+        reader->read(packets[read_packet_index], thisAsUnaryCallback());
+    }
+
+    void processReceivedData()
+    {
+        LOG_TRACE(log, "Received {} packets.", read_packet_index);
+
+        String err_info;
+        if (read_packet_index > 0)
+            has_data = true;
+
+        if (auto error_message = getErrorFromPackets(); unlikely(!error_message.empty()))
+        {
+            closeConnection(fmt::format("Exchange receiver meet error : {}", error_message));
+            return;
+        }
+
+        GRPCReceiveQueueRes res = sendPackets();
+
+        if (!checkResultAfterSendingPackets(res))
+            return;
+
+        startAsyncRead();
+    }
+
+    void processWaitReWrite()
+    {
+        GRPCReceiveQueueRes res = reSendPackets();
+
+        if (!checkResultAfterSendingPackets(res))
+            return;
+
+        stage = AsyncRequestStage::WAIT_BATCH_READ;
+        startAsyncRead();
+    }
+
+    // return true if we still need to receive data from server.
+    bool checkResultAfterSendingPackets(GRPCReceiveQueueRes res)
+    {
+        if (unlikely(sendPacketHasError(res)))
+        {
+            closeConnection("Exchange receiver meet error : push packets fail");
+            return false;
+        }
+
+        if (unlikely(isChannelFull(res)))
+        {
+            asyncWaitForRewrite();
+            return false;
+        }
+
+        if (read_packet_index < batch_packet_count)
+        {
+            stage = AsyncRequestStage::WAIT_FINISH;
+            reader->finish(finish_status, thisAsUnaryCallback());
+            return false;
+        }
+
+        return true;
+    }
+
+    String getErrorFromPackets()
+    {
+        // step 1: check if there is error packet
+        // only the last packet may has error, see execute().
+        if (read_packet_index != 0 && packets[read_packet_index - 1]->hasError())
+            return packets[read_packet_index - 1]->error();
+        // step 2: check memory overflow error
+        for (Int32 i = 0; i < read_packet_index; ++i)
+        {
+            packets[i]->recomputeTrackedMem();
+            if (packets[i]->hasError())
+                return packets[i]->error();
+        }
+        return "";
+    }
+
+    bool retriable() const
+    {
+        return !has_data && retry_times + 1 < max_retry_times;
+    }
+
+    void closeConnection(String && msg)
+    {
+        if (!is_close_conn_called)
+        {
+            stage = AsyncRequestStage::FINISHED;
+            close_conn(!msg.empty(), msg, log);
+            is_close_conn_called = true;
+        }
+    }
+
+    void start()
+    {
+        stage = AsyncRequestStage::WAIT_MAKE_READER;
+
+        // Use lock to ensure async reader is unreachable from grpc thread before this function returns
+        std::lock_guard lock(mu);
+        rpc_context->makeAsyncReader(request, reader, cq, thisAsUnaryCallback());
```

Comment:
I'm a bit confused, will `execute` be called before `makeAsyncReader` returns to callback?

Reply:
> I'm a bit confused, will `execute` be called before `makeAsyncReader` returns to callback?

Yes, execute may be called when reader in the init() is being assigned.

https://github.com/pingcap/tiflash/blob/f9d16c46e0ed6b66494d764954641d20cd4efb19/dbms/src/Flash/Mpp/GRPCReceiverContext.cpp#L105-L113